Properly cleans up routes [prod-leak-fix]

chunhtai opened the pull request:
fixes https://github.com/flutter/flutter/issues/126100

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

chunhtai committed:
Properly cleans up routes

chunhtai commented on packages/flutter/lib/src/widgets/heroes.dart:
```diff
@@ -849,40 +849,48 @@ class HeroController extends NavigatorObserver {
-    if (toRoute != fromRoute && toRoute is PageRoute<dynamic> && fromRoute is PageRoute<dynamic>) {
-      final PageRoute<dynamic> from = fromRoute;
-      final PageRoute<dynamic> to = toRoute;
+    if (toRoute == fromRoute) {
```
this just refactor the if condition to return early instead of nested condition.

the real functional change is in line 889 where it check for whether the route is still active before actually do the fly.

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -3073,7 +3073,16 @@ class _RouteEntry extends RouteTransitionRecord {
+            //
+            // Use a microtask to ensure the overlayEntries have finished
+            // notifying their listeners before disposing.
+            scheduleMicrotask(() {
```
I am actually run into a dead end here.

This microtask causes issue where navigator.dispose expects all route are disposed. I think we may need a way to clean up the entry out of term while destroying all the resource other than the entry's changenotifier.

goderbauer commented:
Looks like many checks are failing. is that related to this change?

goderbauer commented on packages/flutter/lib/src/widgets/heroes.dart:
```diff
@@ -849,40 +849,48 @@ class HeroController extends NavigatorObserver {
+    if (toRoute == fromRoute) {
+      return;
+    }
+    if (toRoute is! PageRoute<dynamic> || fromRoute is! PageRoute<dynamic>) {
```
can you combine this with the previous if?

goderbauer commented on packages/flutter/lib/src/widgets/overlay.dart:
```diff
@@ -214,7 +214,9 @@ class OverlayEntry implements Listenable {
   void dispose() {
     assert(!_disposedByOwner);
-    assert(_overlay == null, 'An OverlayEntry must first be removed from the Overlay before dispose is called.');
+    // In case the owner of this Overlay
```
this comment seems to be incomplete?

chunhtai replied:
sorry about this, I should have turn this into draft. The mounted state was because if the owner of this overlay entry is above the overlay widget. It is possible the entry will be disposed without the owner to have a chance to remove the entry. This is because the dispose starts from the child.

what the parent would know is the entry's notify the parent it has been disposed.

Things can be even worse if the entry is under another opaque entry and does not maintain state. In this case, owner won't even know the entry is disposed because it was never built

chunhtai replied:
never mind If i try really hard, i can still solve this problem

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -3596,6 +3605,9 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
     _updateHeroController(null);
     focusNode.dispose();
     for (final _RouteEntry entry in _history) {
+      // for (final OverlayEntry overlayEntry in entry.route.overlayEntries) {
```
What about this commented out code?

chunhtai commented on packages/flutter/test/widgets/routes_test.dart:
```diff
@@ -401,7 +401,7 @@ void main() {
     );
     await tester.pumpWidget(Container());
-    expect(results, equals(<String>['A: dispose', 'b: dispose']));
+    expect(results, equals(<String>['b: dispose', 'A: dispose']));
```
I think this is more correct as it disposed the top most route first.

chunhtai committed:
typo

chunhtai commented:
Several things makes this pr difficult than I think.

Some systems like HeroController have been operated on disposed Route, and they were somehow able to do that because we didn't properly clean up the route

There are also some logic to make sure the Route's dispose needs to wait for its subtree is properly disposed. This because the subtree may be using something like route's animationcontroller.

Therefore a notifier is introduced in overlayentry to let owner to listen for subtree disposal, but the problem with such logic is that you can't dispose the overlayentry when the notifer notify about the subtree disposal because of the changenotifier doesn't allow calling dispose during listener call back. This was previous ok because the route dispose doesn't properly clean up the overlay entry. Thus, I added a microtask to make sure the listener callback is done before disposing the route.

This created another issue that if a navigator is disposed, it expects all route to dispose. Because some of the dispose was scheduled in a microtask, the navigator won't be able to dispose. Therefore, I added a new lifecycle state called disposing. Navigator will keep track of routes in this state and will force these routes to dispose if Navigator itself is about to be disposed. This is ok because we know the subtree of the route must have been disposed already and we can safely clean up the route at this point

chunhtai committed:
fix comments

chunhtai commented:
a friendly bump

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -2821,6 +2822,8 @@ enum _RouteLifecycle {
   removing, // we are waiting for subsequent routes to be done animating, then will switch to dispose
   // routes that are completely removed from the navigator and overlay.
   dispose, // we will dispose the route momentarily
+  disposing, // The entry is waiting for its subtree to disposed first. It should
+             // be in stored in the _entryWaitingForSubTreeDisposal
```
subtree = widget subtree? Maybe a little clearer:

"The entry is waiting for its widget subtree to be disposed first. It is stored in _entryWaitingForSubTreeDisposal while awaiting that."

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -3047,9 +3050,15 @@ class _RouteEntry extends RouteTransitionRecord {
+    route.dispose();
+  }
+
+  void dispose() {
```
Maybe leave a comment here or on forcedDispose explaining the difference...

goderbauer replied:
e.g. "dispose" waits for all overlayEntries to be unmounted before disposing the route.

goderbauer reviewed (approved):
LGTM

chunhtai committed:
updatE
